receiver/postgresqlreceiver/scraper.go

```diff
@@ -0,0 +1,283 @@
+		tls:      p.config.TLSClientSetting,
+		address:  p.config.NetAddr,
+	})
+}
+
+// initMetric initializes a metric with a metadata attribute.
+func initMetric(ms pdata.MetricSlice, mi metadata.MetricIntf) pdata.Metric {
+	m := ms.AppendEmpty()
+	mi.Init(m)
+	return m
+}
+
+// addToIntMetric adds and attributes a int sum datapoint to metricslice.
+func addToIntMetric(metric pdata.NumberDataPointSlice, attributes pdata.AttributeMap, value int64, ts pdata.Timestamp) {
+	dataPoint := metric.AppendEmpty()
+	dataPoint.SetTimestamp(ts)
+	dataPoint.SetIntVal(value)
+	if attributes.Len() > 0 {
+		attributes.CopyTo(dataPoint.Attributes())
+	}
+}
+
+// scrape scrapes the metric stats, transforms them and attributes them into a metric slices.
+func (p *postgreSQLScraper) scrape(context.Context) (pdata.Metrics, error) {
+	// metric initialization
+	md := pdata.NewMetrics()
+	ilm := md.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otel/postgresql")
+	now := pdata.NewTimestampFromTime(time.Now())
+
+	blocksRead := initMetric(ilm.Metrics(), metadata.M.PostgresqlBlocksRead).Sum().DataPoints()
+	commits := initMetric(ilm.Metrics(), metadata.M.PostgresqlCommits).Sum().DataPoints()
+	databaseSize := initMetric(ilm.Metrics(), metadata.M.PostgresqlDbSize).Gauge().DataPoints()
+	backends := initMetric(ilm.Metrics(), metadata.M.PostgresqlBackends).Gauge().DataPoints()
+	databaseRows := initMetric(ilm.Metrics(), metadata.M.PostgresqlRows).Gauge().DataPoints()
+	operations := initMetric(ilm.Metrics(), metadata.M.PostgresqlOperations).Sum().DataPoints()
+	rollbacks := initMetric(ilm.Metrics(), metadata.M.PostgresqlRollbacks).Sum().DataPoints()
+
+	databaseAgnosticMetricsCollected := false
+	databases := p.config.Databases
+	if len(databases) == 0 {
+		listClient, err := initializeClient(p, "")
+		if err != nil {
+			p.logger.Error("Failed to initialize connection to postgres", zap.Error(err))
+			return md, err
+		}
+		defer listClient.Close()
+
+		dbList, err := listClient.listDatabases()
+		if err != nil {
+			p.logger.Error("Failed to request list of databases from postgres", zap.Error(err))
+			return md, err
+		}
+
+		databases = dbList
+		p.databaseAgnosticMetricCollection(
+			now,
+			listClient,
+			databases,
+			commits,
+			rollbacks,
+			databaseSize,
+			backends,
+		)
+		databaseAgnosticMetricsCollected = true
+	}
+
+	for _, database := range databases {
+		dbClient, err := initializeClient(p, database)
+		if err != nil {
+			// TODO - do string interpolation better
+			p.logger.Error("Failed to initialize connection to postgres", zap.String("database", database), zap.Error(err))
+			continue
+		}
+
+		defer dbClient.Close()
+
+		if !databaseAgnosticMetricsCollected {
+			p.databaseAgnosticMetricCollection(
+				now,
+				dbClient,
+				databases,
+				commits,
+				rollbacks,
+				databaseSize,
+				backends,
+			)
+			databaseAgnosticMetricsCollected = true
+		}
```

Comment:
Does this need to be scoped within this loop, or even within the "list databases" section above? Is it just for the benefit of reusing a client? If so, it seems like the readability benefit of isolating this would be worth the extra client.

Reply:
Maybe something like this

```go
	databases := p.config.Databases
	instanceClient := initializeClient(p, "")
	defer instanceClient.Close()

	if len(databases) == 0 {
		if err != nil {
			p.logger.Error("Failed to initialize connection to postgres", zap.Error(err))
			return md, err
		}

		dbList, err := instanceClient.listDatabases()
		if err != nil {
			p.logger.Error("Failed to request list of databases from postgres", zap.Error(err))
			return md, err
		}

		databases = dbList
	}

	p.databaseAgnosticMetricCollection(
		now,
		instanceClient,
		databases,
		commits,
		rollbacks,
		databaseSize,
		backends,
	)

	for _, database := range databases {
		dbClient, err := initializeClient(p, database)
		if err != nil {
			p.logger.Error("Failed to initialize connection to postgres", zap.String("database", database), zap.Error(err))
			continue
		}
		defer dbClient.Close()

		p.databaseSpecificMetricCollection(
			now,
			dbClient,
			blocksRead,
			databaseRows,
			operations,
		)
	}

	return md, nil
```